### ISSUE TITLE
Array access on struct with array as the single member confuses the analyser

### ISSUE THREAD

**Issue by VisualEhrmanntraut** [Vector35/binaryninja-api#5820]
**Version and Platform (required):**
- Binary Ninja Version: 4.2.5821-dev
- OS: macOS
- OS Version: 15.1 Beta (24B5009l)
- CPU Architecture: Apple Silicon (M3)

**Bug Description:**
Array access using complex math confuses the analyser. Not sure if I've defined something wrong, I'm fairly confident that my type is correct, it seems to be an 0x18-sized structure array inside of a larger structure. The math it does ends up with `allocType * 0x18` anyway.

**Steps To Reproduce:**
Not sure, maybe the disassembly gives you a clue.

**Expected Behavior:**
```c
return (&pAllocations->entries[0]:0x10.q)[zx.q(allocType) * 3]
```

should be
```c
return pAllocations->entries[allocType].ptr;
```

**Screenshots/Video Recording:**
Disassembly
<img width="819" alt="Screenshot 2024-08-03 at 18 56 15" src="https://github.com/user-attachments/assets/b9db5e93-c29c-41ec-bede-85e8eef93f4d">

LLIL
<img width="813" alt="Screenshot 2024-08-03 at 18 56 26" src="https://github.com/user-attachments/assets/adef3712-1173-4961-b401-afb617912895">

MLIL
<img width="825" alt="Screenshot 2024-08-03 at 18 56 35" src="https://github.com/user-attachments/assets/02f15560-38b7-483f-99d2-93f64acd29e1">

HLIL
<img width="841" alt="Screenshot 2024-08-03 at 18 56 42" src="https://github.com/user-attachments/assets/c12d8513-7dab-4ccf-8b10-6c8eb42d1e30">

**Comment by negasora** [comment]
Are you able to share the .bndb?

**Comment by VisualEhrmanntraut** [comment]
@negasora I can share the type definitions. Here they are:

```
enum AMD_TTL_ALLOCATION_TYPE : uint32_t
{
    ALLOCATION_TYPE_EXT_SVCS = 0x1,
    ALLOCATION_TYPE_IPI = 0x4,
    ALLOCATION_TYPE_TLS = 0x5,
    ALLOCATION_TYPE_CGS_DEVICE = 0x6,
    ALLOCATION_TYPE_MAX = 0x16
};

struct AMD_TTL_ALLOCATION_ENTRY
{
    enum AMD_TTL_ALLOCATION_TYPE allocType;
    uint64_t length;
    void* ptr;
};

struct AMD_TTL_ALLOCATIONS
{
    struct AMD_TTL_ALLOCATION_ENTRY entries[0x16];
};
```

**Comment by VisualEhrmanntraut** [comment]
Seems like it gets fixed if I do this
<img width="857" alt="Screenshot 2024-08-03 at 22 31 55" src="https://github.com/user-attachments/assets/e476d93a-45fc-48ae-9e6c-385a2def7523">

It looks like the problem is that if it's a pointer to a structure with an array at offset 0 it will get confused (?)

**Comment by plafosse** [comment]
@VisualEhrmanntraut what did you do in the above image?

**Comment by VisualEhrmanntraut** [comment]
Used a pointer to the type used in the array inside of the structure. Compare the HLIL image to the image you are referring to and look at the type definitions to understand further.

**Comment by plafosse** [comment]
ah yeah I see it now thanks

**Comment by D0ntPanic** [comment]
Fixed in build 4.2.6351
